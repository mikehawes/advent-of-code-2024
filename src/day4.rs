mod find_cursor;
mod lines;
pub mod word_search;
