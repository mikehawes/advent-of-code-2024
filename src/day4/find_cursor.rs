pub struct FindCursor {
    find: Vec<char>,
    find_index: usize,
}

impl FindCursor {
    pub fn start(string: &str) -> FindCursor {
        FindCursor {
            find: string.chars().collect(),
            find_index: 0,
        }
    }
    pub fn char(&self) -> Option<char> {
        self.find.get(self.find_index).copied()
    }
    pub fn advance(&mut self) {
        self.find_index += 1;
    }
    pub fn reset(&mut self) {
        self.find_index = 0;
    }
}

impl Iterator for FindCursor {
    type Item = char;

    fn next(&mut self) -> Option<Self::Item> {
        let char = self.char();
        self.advance();
        char
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn can_iterate_over_chars() {
        let letters: Vec<char> = FindCursor::start("XMAS").collect();
        assert_eq!(letters, vec!['X', 'M', 'A', 'S'])
    }

    #[test]
    fn can_reset() {
        let mut cursor = FindCursor::start("AB");
        let first = vec![cursor.next(), cursor.next(), cursor.next()];
        cursor.reset();
        let second = vec![cursor.next(), cursor.next(), cursor.next()];

        assert_eq!(first, vec![Some('A'), Some('B'), None]);
        assert_eq!(second, vec![Some('A'), Some('B'), None]);
    }
}
