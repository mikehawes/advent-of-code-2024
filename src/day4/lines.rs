use std::cmp::min;

pub type Point = (usize, usize);
pub type Line = Vec<Point>;
pub type LinesIter = dyn Iterator<Item = Line>;

pub fn generate_lines(width: usize, height: usize) -> Box<LinesIter> {
    Box::new(
        generate_horizontals(width, height)
            .chain(generate_verticals(width, height))
            .chain(generate_diagonals(width, height)),
    )
}
fn generate_horizontals(width: usize, height: usize) -> Box<LinesIter> {
    let starts_at_left = 0..height;
    Box::new(starts_at_left.map(move |y| horizontal_from_left(width, y)))
}
fn horizontal_from_left(width: usize, y: usize) -> Line {
    let progress = 0..width;
    progress.clone().map(move |x| (x, y)).collect()
}
fn generate_verticals(width: usize, height: usize) -> Box<LinesIter> {
    let starts_at_top = 0..width;
    Box::new(starts_at_top.map(move |x| vertical_from_top(x, height)))
}
fn vertical_from_top(x: usize, height: usize) -> Line {
    let progress = 0..height;
    progress.map(move |y| (x, y)).collect()
}
fn generate_diagonals(width: usize, height: usize) -> Box<LinesIter> {
    let down_right = generate_diagonals_down_right(width, height);
    let up_right = generate_diagonals_up_right(width, height);
    Box::new(down_right.chain(up_right))
}

fn generate_diagonals_down_right(width: usize, height: usize) -> Box<LinesIter> {
    let starts_at_side = (1..(height - 1)).rev();
    let starts_at_top = 0..(width - 1);
    Box::new(
        starts_at_side
            .map(move |y| diagonal_from_side_down(width, height, y))
            .chain(starts_at_top.map(move |x| diagonal_from_top(width, height, x))),
    )
}

fn generate_diagonals_up_right(width: usize, height: usize) -> Box<LinesIter> {
    let starts_at_side = 1..(height - 1);
    let starts_at_bottom = 0..(width - 1);
    Box::new(
        starts_at_side
            .map(move |y| diagonal_from_side_up(width, y))
            .chain(starts_at_bottom.map(move |x| diagonal_from_bottom(width, height, x))),
    )
}

fn diagonal_from_top(width: usize, height: usize, x: usize) -> Line {
    let diffs = 0..min(height, width - x);
    diffs.map(move |diff| (x + diff, diff)).collect()
}

fn diagonal_from_bottom(width: usize, height: usize, x: usize) -> Line {
    let diffs = 0..min(height, width - x);
    diffs
        .map(move |diff| (x + diff, height - diff - 1))
        .collect()
}

fn diagonal_from_side_down(width: usize, height: usize, y: usize) -> Line {
    let diffs = 0..min(width, height - y);
    diffs.map(move |diff| (diff, y + diff)).collect()
}

fn diagonal_from_side_up(width: usize, y: usize) -> Line {
    let diffs = 0..min(width, y + 1);
    diffs.map(move |diff| (diff, y - diff)).collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use std::collections::HashMap;
    use std::fmt::{Error, Write};

    #[test]
    fn can_generate_2_by_2_lines() {
        assert_snapshot!(print_lines(2, 2).unwrap())
    }

    #[test]
    fn can_generate_3_by_3_lines() {
        assert_snapshot!(print_lines(3, 3).unwrap())
    }

    #[test]
    fn can_generate_4_by_4_lines() {
        assert_snapshot!(print_lines(4, 4).unwrap())
    }

    #[test]
    fn can_generate_10_by_10_lines() {
        assert_snapshot!(print_lines(10, 10).unwrap())
    }

    fn print_lines(width: usize, height: usize) -> Result<String, Error> {
        let mut str: String = "".to_owned();
        print_lines_iter(generate_lines(width, height), width, height, &mut str)?;
        Ok(str)
    }

    fn print_lines_iter<W: Write>(
        lines: Box<LinesIter>,
        width: usize,
        height: usize,
        out: &mut W,
    ) -> std::fmt::Result {
        for line in lines {
            print_line(line, width, height, out)?;
            out.write_char('\n')?
        }
        Ok(())
    }

    fn print_line<W: Write>(
        line: Line,
        width: usize,
        height: usize,
        out: &mut W,
    ) -> std::fmt::Result {
        let point_to_n: HashMap<(usize, usize), usize> =
            HashMap::from_iter(line.iter().enumerate().map(|(n, point)| (*point, n)));
        for y in 0..height {
            for x in 0..width {
                if let Some(n) = point_to_n.get(&(x, y)) {
                    out.write_char(char::from_digit(((n + 1) % 10) as u32, 10).unwrap())?;
                } else {
                    out.write_char('.')?;
                }
            }
            out.write_char('\n')?;
        }
        Ok(())
    }
}
