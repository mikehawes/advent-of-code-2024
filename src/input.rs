use std::fs::read_to_string;
use std::io;
use std::path::{Path, PathBuf};

pub fn input_to_string<P: AsRef<Path>>(path: P) -> io::Result<String> {
    let full_path = file_in_src(path);
    read_to_string(full_path)
}

pub fn file_in_src<P: AsRef<Path>>(path: P) -> PathBuf {
    let mut full_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    full_path.push("src");
    full_path.push(path);
    full_path
}
