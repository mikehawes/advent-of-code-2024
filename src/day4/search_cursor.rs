use crate::day4::WordSearch;

pub struct SearchCursor<'a> {
    search: &'a WordSearch,
    x: usize,
    y: usize,
}

impl<'a> SearchCursor<'a> {
    pub fn top_left(search: &'a WordSearch) -> SearchCursor<'a> {
        SearchCursor { search, x: 0, y: 0 }
    }

    pub fn char(&self) -> Option<char> {
        self.search
            .tiles
            .get(self.y)
            .and_then(|line| line.get(self.x).copied())
    }
    pub fn advance_x(&mut self) -> bool {
        if self.x < self.search.width {
            self.x += 1;
            true
        } else {
            false
        }
    }
    pub fn next_x(&mut self) -> Option<char> {
        let char = self.char();
        self.advance_x();
        char
    }
    pub fn reset_x(&mut self) {
        self.x = 0;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn can_read_top_left() {
        let word_search = WordSearch::parse("ABCD");
        let cursor = SearchCursor::top_left(&word_search);
        assert_eq!(cursor.char(), Some('A'))
    }

    #[test]
    fn can_read_row() {
        let word_search = WordSearch::parse("AB");
        let mut cursor = SearchCursor::top_left(&word_search);
        let output = vec![
            cursor.next_x(),
            cursor.next_x(),
            cursor.next_x(),
            cursor.next_x(),
        ];
        assert_eq!(output, vec![Some('A'), Some('B'), None, None])
    }

    #[test]
    fn can_reset_row() {
        let word_search = WordSearch::parse("AB");
        let mut cursor = SearchCursor::top_left(&word_search);
        let first = vec![cursor.next_x(), cursor.next_x(), cursor.next_x()];
        cursor.reset_x();
        let second = vec![cursor.next_x(), cursor.next_x(), cursor.next_x()];
        assert_eq!(first, vec![Some('A'), Some('B'), None]);
        assert_eq!(second, vec![Some('A'), Some('B'), None]);
    }
}
