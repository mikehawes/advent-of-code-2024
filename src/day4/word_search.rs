use crate::day4::find_cursor::FindCursor;
use crate::day4::lines::{generate_lines, Point};
use std::collections::HashSet;

#[derive(Debug, Eq, PartialEq)]
pub struct WordSearch {
    tiles: Vec<Vec<char>>,
    width: usize,
}

impl WordSearch {
    pub fn parse(input: &str) -> WordSearch {
        let tiles: Vec<Vec<char>> = input.lines().map(read_chars).collect();
        let width = tiles.first().map(Vec::len).unwrap_or(0);
        WordSearch { tiles, width }
    }

    pub fn count_xmas(&self) -> i32 {
        let (matches, _) = self.count_xmas_find_points();
        matches
    }

    fn count_xmas_find_points(&self) -> (i32, HashSet<Point>) {
        let mut matches = 0;
        let mut find_cursor = FindCursor::start("XMAS");
        let mut points = HashSet::new();
        let mut current_points = HashSet::new();
        for line in generate_lines(self.width, self.tiles.len()) {
            for point in &line {
                if find_cursor.check_match_advance(self.char(point)) {
                    current_points.insert(*point);
                    if find_cursor.reset_if_finished() {
                        matches += 1;
                        current_points.iter().for_each(|p| {
                            points.insert(*p);
                        });
                        current_points.clear();
                    }
                } else {
                    current_points.clear();
                }
            }
            current_points.clear();
            find_cursor.reset();
            for point in line.iter().rev() {
                if find_cursor.check_match_advance(self.char(point)) {
                    current_points.insert(*point);
                    if find_cursor.reset_if_finished() {
                        matches += 1;
                        current_points.iter().for_each(|p| {
                            points.insert(*p);
                        });
                        current_points.clear();
                    }
                } else {
                    current_points.clear();
                }
            }
            find_cursor.reset();
        }
        (matches, points)
    }

    fn char(&self, point: &Point) -> char {
        let (x, y) = *point;
        self.tiles[y][x]
    }
}

fn read_chars(line: &str) -> Vec<char> {
    line.chars().collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::input::input_to_string;
    use insta::assert_snapshot;
    use std::fmt::Write;

    #[test]
    fn can_find_no_xmas() {
        let word_search = WordSearch::parse("ABCD");
        assert_eq!(word_search.count_xmas(), 0)
    }

    #[test]
    fn can_find_xmas_one_line() {
        let word_search = WordSearch::parse("XMAS");
        assert_eq!(word_search.count_xmas(), 1)
    }

    #[test]
    fn can_find_xmas_two_lines() {
        let word_search = WordSearch::parse("XMAS\nXMAS");
        assert_eq!(word_search.count_xmas(), 2)
    }

    #[test]
    fn can_find_xmas_one_of_two_lines() {
        let word_search = WordSearch::parse("XMAS\nXMEN");
        assert_eq!(word_search.count_xmas(), 1)
    }

    #[test]
    fn can_find_xmas_over_lines() {
        let word_search = WordSearch::parse("X\nM\nA\nS");
        assert_eq!(word_search.count_xmas(), 1)
    }

    #[test]
    fn can_find_xmas_after_x() {
        let word_search = WordSearch::parse("XXMAS");
        assert_eq!(word_search.count_xmas(), 1)
    }

    #[test]
    fn can_find_xmas_in_example() {
        let example = input_to_string("day4/example.txt").unwrap();
        let word_search = WordSearch::parse(example.as_str());
        let (matches, points) = word_search.count_xmas_find_points();
        assert_eq!(matches, 18);
        assert_snapshot!(print_relevant_points(&word_search, &points))
    }

    #[test]
    fn can_find_xmas_diagonal() {
        let word_search = WordSearch::parse(
            "X___\n\
                   _M__\n\
                   __A_\n\
                   ___S",
        );
        assert_eq!(word_search.count_xmas(), 1)
    }

    #[test]
    fn can_parse_tiles() {
        let input = "\
            ABC\n\
            DEF\n\
            GEH";
        assert_eq!(
            WordSearch::parse(input),
            WordSearch {
                tiles: vec![
                    vec!['A', 'B', 'C'],
                    vec!['D', 'E', 'F'],
                    vec!['G', 'E', 'H']
                ],
                width: 3
            }
        );
    }
    fn print_relevant_points(word_search: &WordSearch, points: &HashSet<Point>) -> String {
        let mut str: String = "".to_owned();
        write_relevant_points(word_search, points, &mut str).unwrap();
        str
    }
    fn write_relevant_points<W: Write>(
        word_search: &WordSearch,
        points: &HashSet<Point>,
        out: &mut W,
    ) -> std::fmt::Result {
        for y in 0..word_search.tiles.len() {
            for x in 0..word_search.width {
                if points.contains(&(x, y)) {
                    out.write_char(word_search.char(&(x, y)))?;
                } else {
                    out.write_char('.')?;
                }
            }
            out.write_char('\n')?;
        }
        Ok(())
    }
}
