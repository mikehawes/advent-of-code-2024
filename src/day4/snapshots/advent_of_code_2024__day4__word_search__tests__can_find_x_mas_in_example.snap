---
source: src/day4/word_search.rs
expression: "print_relevant_points(&word_search, &points)"
snapshot_kind: text
---
.M.S......
..A..MSMS.
.M.S.MAA..
..A.ASMSM.
.M.S.M....
..........
S.S.S.S.S.
.A.A.A.A..
M.M.M.M.M.
..........
